rfcs/20201105-attribute-num-threads/README.md

```diff
@@ -0,0 +1,271 @@
+for problem decomposition.
+
+<details>
+<summary>C API</summary>
+
+~~~cpp
+dnnl_status_t DNNL_API dnnl_threadpool_interop_primitive_attr_create(dnnl_primitive_attr_t *attr, uint8_t num_threads);
+~~~
+
+</details>
+
+<details>
+<summary>C++ API</summary>
+
+~~~cpp
+namespace dnnl {
+namespace threadpool_interop {
+
+inline dnnl::primitive_attr make_primitive_attr(uint8_t num_threads = DNNL_MAX_THREADS) {
+    dnnl_primitive_attr_t c_attr;
+    dnnl::error::wrap_c_api(dnnl_threadpool_interop_primitive_attr_create(&c_attr, num_threads));
+}
+
+} // namespace threadpool_interop
+} // namespace dnnl
+~~~
+
+</details>
+
+The `num_threads` value in the primitive attributes will indicate the upper
+limit for the number of threads used for decomposition in the primitive
+(descriptor). The default wildcard value `DNNL_MAX_THREADS` indicates that
+`dnnl_get_max_threads()` will be used internally as an upper limit.
+
+By assigning `0` in the API, the user will instruct the library that no
+threadpool will be used to execute the corresponding primitive, therefore no
+decomposition will be performed and the execution will be performed by the
+calling thread.
+
+By assigning `1` or higher value in the API, the user will instruct the library
+the upper limit for the number of threads for decomposition, and, during
+execution, the parallel sections will be executed by the threads in the
+threadpool.
+
+In the context of primitive cache, the number of threads passed by primitive
+attributes will be used to construct the cache key. In means that two primitives
+created using different `num_threads` will both be stored in the cache as
+separate implementations.
+
+Since the implementation of this RFC can target oneDNN v2.1 at the least, it is
+proposed to introduce the new API into `dnnl_threadpool.hpp`, where Threadpool-specific API is introduced.
+
+Pros:
+- Ability to have different problem decompositions for different primitives;
+- Flexibility with respect to user cost model and Threadpool implementation;
+- Minimal changes required in the library;
+
+Cons:
+- Upper limit for number of threads not applicable to other threading runtimes,
+  may cause some inconsistencies in the library code.
+
+### Option 2: new primitive attributes for all CPU threading runtimes
+
+In order to maintain consistency between threading runtimes, this option
+proposes to introduce a generic primitive attribute, that instructs the library
+the upper limit for thread count during primitive (descriptor) creation.
+
+<details>
+<summary>C API</summary>
+
+~~~cpp
+dnnl_primitive_attr_get_num_threads(dnnl_primitive_attr_t attr, uint8_t *num_threads);
+
+dnnl_primitive_attr_set_num_threads(dnnl_primitive_attr_t attr, uint8_t num_threads);
+~~~
+
+</details>
+
+<details>
+<summary>C++ API</summary>
+
+~~~cpp
+struct primitive_attr : public handle<dnnl_primitive_attr_t> {
+    ...
+
+    uint8_t get_num_threads() const {}
```

Comment:
I was wondering if we can do away with the assumption on the bound on number of threads (i.e. 256 here!)

Reply:
Are you referring to the assumption of using `hardware_concurrency` for decomposition when the library is not aware of a TP?

Reply:
Hi @sbogusev 
I was just referring to the choice of data type `uint8_t` in specifying the number of threads here.

Reply:
Hi Alok. Technically we could use `uint32_t` for a higher upper bound, however I am not sure using a number of threads higher than 256 (in a system with a much larger `hardware_concurrency`) will lead to any performance benefits. I guess we can raise it only for simplicity. Will update in the next revision, thank you!